In [ ]:
# This notebook is to conduct consistency checks over the orders dataset

In [ ]:
# Content List
#
# 01. Importing libraries
# 02. Importing data
# 03. Mixed-type column practice
# 04. Missing value treatment practice (Task 1)
# 05. Duplicate treatment practice (Task 1)
# 06. Exporting checked dataframe (Task 9)
# 07. Additional consistency checks (Exercise Tasks)
# 08. Exporting additional checked dataframe (Task 9)

# 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [4]:
# Defining path
path = r'/Users/bladael/Documents/Learning/CareerFoundry_DA/Data Immersion/Achievement 4/06-2023 Instacart Basket Analysis'

In [8]:
# Import products.csv from Original Data folder
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [9]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [87]:
# Import orders_wrangled.csv from Prepared Data folder
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'))

In [88]:
df_ords.head()

order_id  user_id  order_sequence_number  order_days_of_week  \
0   2539329        1                      1                   2   
1   2398795        1                      2                   3   
2    473747        1                      3                   3   
3   2254736        1                      4                   4   
4    431534        1                      5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# 03. Mixed-type column practice

In [20]:
# Create a dataframe
df_test = pd.DataFrame()

In [21]:
df_test

Empty DataFrame
Columns: []
Index: []

In [22]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [24]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [30]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 160.0+ bytes


In [32]:
# Check for mixed type
# comparing value type in a column against the data type of the first value in the column
# if the data types are not the same, then defined as 'weird' variable
# then if there are more than 0 'weird; variable, print the column heading
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [34]:
# Additional practice - adding one more mixed column to test
df_test['mix_2'] = [123,'haha', False ,'house']

In [35]:
df_test.head()

mix  mix_2
0     a    123
1     b   haha
2     1  False
3  True  house

In [36]:
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix
mix_2


In [37]:
df_test['mix'].value_counts.astype('str')

AttributeError: 'function' object has no attribute 'astype'

In [41]:
# Counting the values of the different data types within a column (mix)
df_test['mix'].apply(type).value_counts()

<class 'str'>     2
<class 'int'>     1
<class 'bool'>    1
Name: mix, dtype: int64

# 04. Missing value treatment practice (Task 1)

In [43]:
# Find null values in a dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [47]:
# Create a new dataframe with null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [48]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [49]:
df_prods.shape

(49693, 5)

In [50]:
# Create a new product dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

# Other alternatives (these would overwrite the existing dataframe)
# Option 1: df_prods.dropna(inplace = True)
# Option 2: df_prods.dropna(subset = ['product_name'], inplace = True)

In [51]:
df_prods_clean.shape

(49677, 5)

# 05. Duplicate treatment practice (Task 1)

In [52]:
# Create a new dataframe containing duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [53]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [54]:
# Check rows before removing duplicate
df_prods_clean.shape

(49677, 5)

In [56]:
# Create a new dataframe without any duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [57]:
# Check rows after removing duplicate
df_prods_clean_no_dups.shape

(49672, 5)

# 06. Exporting checked dataframe (Task 9)

In [58]:
df_prods_clean_no_dups.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [59]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index = False)

# 07. Additional consistency checks (Exercise Tasks)

In [89]:
# Task 2: descriptive stats over df_ords dataframe
df_ords.describe()

order_id       user_id  order_sequence_number  order_days_of_week  \
count  3.421083e+06  3.421083e+06           3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05           1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04           1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00           1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04           5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05           1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05           2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05           1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [90]:
# Task 3: check for mixed-type data
for col in df_ords.columns.tolist():
    weird_ords = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird_ords]) > 0:
        print(col)
    else:
        print(col,': no mixed-type data')

order_id : no mixed-type data
user_id : no mixed-type data
order_sequence_number : no mixed-type data
order_days_of_week : no mixed-type data
order_hour_of_day : no mixed-type data
days_since_prior_order : no mixed-type data


Task 4: Update any mix-type data. 
CP: From the Task 3 above, no columns with mixed data types were noted

In [91]:
# Task 5: Check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_sequence_number          0
order_days_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

CP: 'days_since_prior_order' column is the only column with missing values. It has the 206,209 missing values and I am suspecting these are all first time orders where there were no prior orders

In [92]:
# Task 6: Addressing missing values
# First, checking the assumption noted above by creating a new dataframe with missing values
df_ords_mv = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [93]:
df_ords_mv

order_id  user_id  order_sequence_number  order_days_of_week  \
0         2539329        1                      1                   2   
11        2168274        2                      1                   2   
26        1374495        3                      1                   1   
39        3343014        4                      1                   6   
45        2717275        5                      1                   3   
...           ...      ...                    ...                 ...   
3420930    969311   206205                      1                   4   
3420934   3189322   206206                      1                   3   
3421002   2166133   206207                      1                   6   
3421019   2227043   206208                      1                   1   
3421069   3154581   206209                      1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [94]:
# Task 6: Addressing missing values
# Second, need to verify whether there missing values with the order_sequence_number value not equal to 1
df_ords_mv[df_ords_mv['order_sequence_number'] != 1]

Empty DataFrame
Columns: [order_id, user_id, order_sequence_number, order_days_of_week, order_hour_of_day, days_since_prior_order]
Index: []

CP: Based on my analysis above, it is confirmed that 'days_since_prior order' missing values are due to the fact that they were all first time orders based on their 'order_sequence_numbers.' The missing values will be all replaced with in integer 0.

In [95]:
df_ords.shape

(3421083, 6)

In [96]:
# Task 6: Address missing values
# Replacing missing values to 0
df_ords['days_since_prior_order'].fillna(0, inplace = True)

In [97]:
df_ords.shape

(3421083, 6)

In [98]:
df_ords.isnull().sum()

order_id                  0
user_id                   0
order_sequence_number     0
order_days_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [99]:
df_ords.head()

order_id  user_id  order_sequence_number  order_days_of_week  \
0   2539329        1                      1                   2   
1   2398795        1                      2                   3   
2    473747        1                      3                   3   
3   2254736        1                      4                   4   
4    431534        1                      5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     0.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [103]:
# Task 7: Check duplidates
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, order_sequence_number, order_days_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Task 8: CP: no duplicates were found from the step above

# 08. Exporting additional checked dataframe (Task 9)

In [104]:
df_ords.head()

order_id  user_id  order_sequence_number  order_days_of_week  \
0   2539329        1                      1                   2   
1   2398795        1                      2                   3   
2    473747        1                      3                   3   
3   2254736        1                      4                   4   
4    431534        1                      5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     0.0  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [105]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index = False)